clients/src/main/java/org/apache/kafka/clients/producer/internals/DefaultPartitioner.java

```diff
@@ -52,36 +50,38 @@ public void configure(Map<String, ?> configs) {}
      */
     public int partition(String topic, Object key, byte[] keyBytes, Object value, byte[] valueBytes, Cluster cluster) {
-        List<PartitionInfo> partitions = cluster.partitionsForTopic(topic);
-        int numPartitions = partitions.size();
-        if (keyBytes == null) {
-            int nextValue = nextValue(topic);
-            List<PartitionInfo> availablePartitions = cluster.availablePartitionsForTopic(topic);
-            if (availablePartitions.size() > 0) {
-                int part = Utils.toPositive(nextValue) % availablePartitions.size();
-                return availablePartitions.get(part).partition();
-            } else {
-                // no partitions are available, give a non-available partition
-                return Utils.toPositive(nextValue) % numPartitions;
-            }
-        } else {
-            // hash the keyBytes to choose a partition
-            return Utils.toPositive(Utils.murmur2(keyBytes)) % numPartitions;
-        }
+        throw new UnsupportedOperationException("DefaultPartitioner no longer supports the partition method.");
     }
 
-    private int nextValue(String topic) {
-        AtomicInteger counter = topicCounterMap.get(topic);
-        if (null == counter) {
-            counter = new AtomicInteger(ThreadLocalRandom.current().nextInt());
-            AtomicInteger currentCounter = topicCounterMap.putIfAbsent(topic, counter);
-            if (currentCounter != null) {
-                counter = currentCounter;
+    public void close() {}
+    
+    /**
+     * Computes a partition and based on relevant data. If the record does not have a key, will either return the
+     * current sticky partition or switch partitions if we are creating a new batch. If the record has a key,
+     * simply return the partition calculated by the hashed value.
+     * 
+     * @param topic The topic name
+     * @param key The key to partition on (or null if no key)
+     * @param keyBytes The serialized key to partition on( or null if no key)
+     * @param value The value to partition on or null
+     * @param valueBytes The serialized value to partition on or null
+     * @param cluster The current cluster metadata
+     * @param isNewBatch A boolean to specify whether a new batch will be created
+     * @return The record's computed partition
+     */
+    public int computePartition(String topic, Object key, byte[] keyBytes, Object value, byte[] valueBytes, Cluster cluster,  boolean isNewBatch) {
+        if (isNewBatch && keyBytes == null) {
+            return stickyPartitioner.nextPartition(topic, cluster);
+        } else if (keyBytes == null) {
+            return stickyPartitioner.partition(topic, cluster);
+        } else {
+            List<PartitionInfo> partitions = cluster.partitionsForTopic(topic);
+            int numPartitions = partitions.size();
+            if (isNewBatch) {
+                stickyPartitioner.nextPartition(topic, cluster);
```

Comment:
Hmm, it this right? If we get here, the record has a key. So, it seems that we should use the hash of the key to determine the partition regardless of the value of isNewBatch?

Reply:
I realize the logic is a little confusing here, and I could probably clean it up. The first case is new batch + no key, the second case is no key, and the third case is keyed. There was a discussion of still changing the sticky partition on keyed values (in the case where we have mixed keys and values, we could potentially get stuck on the same sticky partition). So within this case, there is a nested if to change the partition. However, in any case, we return the hashed key in line 84. (For some reason, github took the bracket out of the diff and I don't know why.)

Reply:
Thanks for the explanation. Make sense.